packages/grafana-ui/src/components/Text/TextElements.tsx

```diff
@@ -1,59 +1,60 @@
-import React from 'react';
+import React, { createElement, forwardRef, useCallback } from 'react';
+import { useStyles2 } from 'src/themes';
 
-import { Text, TextProps } from './Text';
+import { ThemeTypographyVariantTypes } from '@grafana/data';
 
-interface TextElementsProps extends Omit<TextProps, 'as'> {}
+import { getTextStyles, TextStyleProps } from './style';
 
-export const H1 = React.forwardRef<HTMLHeadingElement, TextElementsProps>((props, ref) => {
-  return <Text as="h1" {...props} variant={props.variant || 'h1'} ref={ref} />;
-});
+type TextElement = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'p' | 'legend';
 
-H1.displayName = 'H1';
+interface TextElementProps extends TextStyleProps {
+  children: React.ReactNode;
+}
 
-export const H2 = React.forwardRef<HTMLHeadingElement, TextElementsProps>((props, ref) => {
-  return <Text as="h2" {...props} variant={props.variant || 'h2'} ref={ref} />;
-});
+function createTextElement(element: TextElement, defaultVariant: keyof ThemeTypographyVariantTypes) {
```

Comment:
I find using a function very interesting as we create components based on the HTML tag we want to render and the variant that specifies the styles to be applied.